### ISSUE TITLE
Minor bug in jax.vjp? (difference w.r.t. jax.grad on scalar function)

### ISSUE THREAD

**Issue by dpotop** [google/jax#19479]
### Description

Consider the following function: `f = lambda a : jax.numpy.multiply(a,a)`.
I compute its gradient using not `grad`, but `vjp` (it is something I want, to be able to handle functions with multiple outputs). Given that `grad` is characterized in the doc as a restriction of `vjp`, the result should be the same. And yet, after applying the following code:

```
@jit
def myfun(x):
    _,y = jax.vjp(f,x)
    return y
exchange_dialect = 'stablehlo'
str_hlo = str(myfun.lower(1.0).compiler_ir(dialect=exchange_dialect))
print(str_hlo)

```
I obtain:
```
func.func public @main(%arg0: tensor<f32> {mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) -> (tensor<f32> {jax.result_info = "[<flat index 0>][0][<flat index 0>][0][0]", mhlo.layout_mode = "default"}) {
    return %arg0 : tensor<f32>
  }
```
The result is correct if I use `grad` instead of `vjp`, or if the function `f` is defined as `jax.numpy.square`. This makes me think that the bug is related to the handling of broadcast (the fact that variable `a` is used twice inside the lambda). 

### What jax/jaxlib version are you using?

0.4.23 0.4.23

### Which accelerator(s) are you using?

CPU

### Additional system info?

Python 3.11.4

**Comment by jakevdp** [comment]
What is `f` meant to be in your code snippet?

**Comment by hawkinsp** [comment]
The second value returned by `vjp` is a pullback function. If you return it, from a `jit`-decorated function JAX *will* return it and treat it as opaque, but it would be an error to call it.

See the documentation for an example of how to use it: https://jax.readthedocs.io/en/latest/_autosummary/jax.vjp.html

Does that help?
